lib/utils/genmap/genmap.go

```diff
@@ -0,0 +1,362 @@
+// said values need to be regenerated very frequently, and possibly in response to events.
+type GenMap[K comparable, V any] struct {
+	cfg     Config[K, V]
+	rw      sync.RWMutex
+	entries map[K]*entry[V]
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+// CheckAndSetDefaults verifies config parameters and sets
+// default values as needed.
+func (c *Config[K, V]) CheckAndSetDefaults() error {
+	if c.Context == nil {
+		c.Context = context.Background()
+	}
+
+	if c.Generator == nil {
+		return trace.Wrap(ErrNoGenerator)
+	}
+
+	if c.RegenInterval < 1 {
+		c.RegenInterval = time.Minute
+	}
+
+	if c.Jitter == nil {
+		c.Jitter = retryutils.NewSeventhJitter()
+	}
+
+	if c.MaxFailures < 1 {
+		c.MaxFailures = 3
+	}
+
+	return nil
+}
+
+// New sets up a new GenMap based on the supplied configuration.
+func New[K comparable, V any](cfg Config[K, V]) (*GenMap[K, V], error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(cfg.Context)
+
+	return &GenMap[K, V]{
+		cfg:     cfg,
+		entries: make(map[K]*entry[V]),
+		ctx:     ctx,
+		cancel:  cancel,
+	}, nil
+}
+
+type entry[V any] struct {
+	ptr       atomic.Pointer[value[V]]
+	init      chan struct{}
+	wantRegen chan struct{}
+	ctx       context.Context
+	cancel    context.CancelFunc
+	done      chan struct{}
+}
+
+type value[V any] struct {
+	val        V
+	err        error
+	regenOnErr chan struct{}
+}
+
+func (v *value[V]) read() (V, error) {
+	if v.err != nil {
+		// reads of error values trigger early regen
+		select {
+		case v.regenOnErr <- struct{}{}:
+		default:
+		}
+	}
+	return v.val, v.err
+}
+
+// Generate starts/restarts generation for the target key. This can be called to trigger generation for keys that
+// are not yet being generated, or to trigger and early regen for existing keys.
+func (m *GenMap[K, V]) Generate(key K) {
+	ent, isNew := m.ensureEntry(key)
+	if isNew {
+		return
+	}
+
+	select {
+	case ent.wantRegen <- struct{}{}:
+	default:
+	}
+}
+
+// RegenAll triggers regeneration of all currently tracked values.
+func (m *GenMap[K, V]) RegenAll() {
+	m.rw.RLock()
+	defer m.rw.RUnlock()
+
+	for _, ent := range m.entries {
+		select {
+		case ent.wantRegen <- struct{}{}:
+		default:
+		}
+	}
+}
+
+// Terminate halts generation for the target key. Note that future calls to Get/Generate will restart generation. For this
+// reason a robust 'cleanup' strategy should also include providing a generator that returns errors when called for keys
+// that are no longer needed. This will ensure that concurrent get/generate calls can't result in dangling entries long-term,
+// since the entry will eventually get cleaned up naturally when it hits MaxFailures.
+func (m *GenMap[K, V]) Terminate(key K) {
+	m.rw.RLock()
+	_, ok := m.entries[key]
+	m.rw.RUnlock()
+	if !ok {
+		return
+	}
+
+	m.rw.Lock()
+	ent, ok := m.entries[key]
+	if ok {
+		delete(m.entries, key)
```

Comment:
> For this reason a robust 'cleanup' strategy should also include providing a generator that returns errors when called for keys that are no longer needed. This will ensure that concurrent get/generate calls can't result in dangling entries long-term, since the entry will eventually get cleaned up naturally when it hits MaxFailures.

Maybe we can skip `delete(m.entries, key)` here and rely on the `delete` when `genarate()` finishes? That way concurrent get can get a cache until terminate is complete.

Reply:
Gave this a shot, but ended up reverting it.  The concurrent termination test is a lot more difficult to pull off without flakiness if `Terminate` doesn't remove the entry, and there isn't much benefit to continuing to serve the old value for a few more ms.  If there are concurrent reads, we're almost certainly gonna be reloading the entry anyway.  Removing this does simplify the logic slightly, but I think the the current behavior is a bit easier to test and reason about.